cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,426 @@
+  rmm::device_uvector<order> d_column_order;
+  rmm::device_uvector<null_order> d_null_precedence;
+  rmm::device_uvector<size_type> d_depths;
+  bool _has_nulls;
+};
+
+/**
+ * @brief An owning object that can be used to lexicographically compare two rows of the same table
+ *
+ * This class can take a table_view and preprocess certain columns to allow for lexicographical
+ * comparison. The preprocessed table and temporary data required for the comparison are created and
+ * owned by this class.
+ *
+ * This class can then provide a functor object that can used on the device.
+ * The object of this class must outlive the usage of the device functor.
+ */
+class self_comparator {
+ public:
+  /**
+   * @brief Construct an owning object for performing a lexicographic comparison between two rows of
+   * the same table.
+   *
+   * @param table The table to compare
+   * @param column_order Optional, host array the same length as a row that indicates the desired
+   * ascending/descending order of each column in a row. If empty, it is assumed all columns are
+   * sorted in ascending order.
+   * @param null_precedence Optional, device array the same length as a row and indicates how null
+   * values compare to all other for every column. If empty, then null precedence would be
+   * `null_order::BEFORE` for all columns.
+   * @param stream The stream to construct this object on. Not the stream that will be used for
+   * comparisons using this object.
+   */
+  self_comparator(table_view const& t,
+                  host_span<order const> column_order         = {},
+                  host_span<null_order const> null_precedence = {},
+                  rmm::cuda_stream_view stream                = rmm::cuda_stream_default)
+    : d_t{std::make_shared<preprocessed_table>(t, column_order, null_precedence, stream)}
+  {
+  }
+
+  /**
+   * @brief Construct an owning object for performing a lexicographic comparison between two rows of
+   * the same preprocessed table.
+   *
+   * @param t A table preprocessed for lexicographic comparison
+   */
+  self_comparator(std::shared_ptr<preprocessed_table> t) : d_t{std::move(t)} {}
+
+  /**
+   * @brief Ger the comparison operator to use on the device
+   *
+   * @tparam Nullate Optional, A cudf::nullate type describing how to check for nulls.
+   */
+  template <typename Nullate = nullate::DYNAMIC>
+  device_row_comparator<Nullate> device_comparator()
+  {
+    if constexpr (std::is_same_v<Nullate, nullate::DYNAMIC>) {
+      return device_row_comparator(Nullate{d_t->has_nulls()},
+                                   *d_t,
+                                   *d_t,
+                                   d_t->depths(),
+                                   d_t->column_order(),
+                                   d_t->null_precedence());
+    } else {
+      return device_row_comparator<Nullate>(
+        Nullate{}, *d_t, *d_t, d_t->depths(), d_t->column_order(), d_t->null_precedence());
+    }
```

Comment:
Minor nit: It might look a bit cleaner if you just assigned a temporary Nullate in the `if constexpr` then called the constructor only once. Up to you though.

Reply:
`nullate::DYNAMIC` is the only one that constructs with a bool arg. The other two `nullate::YES/NO` don't take an arg.

Reply:
The caller of `device_comparator` is deciding dynamic or not by virtue of the template argument. So why not make them create the nullate object and pass it in?
```
template <typename Nullate>
  device_row_comparator<Nullate> device_comparator(Nullate nulls) {
    return device_row_comparator(
        nulls, *d_t, *d_t, d_t->depths(), d_t->column_order(), d_t->null_precedence());
}
```